### PR TITLE
refactor(es/helpers): Remove unnecessary exports

### LINKED ISSUE #9203: Helper methods do not need to be export functions
### Describe the feature

`_` export is enough, and we can rename them while exporting.

### Babel plugin or link to the feature description

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by devjiwonchoi**
**Description:**

Helper methods were exporting twice, `_` and it's own function name.
We can map the build script to export `_` as it's own name (filename).
Therefore we export only once as `_`, and map the name on the build script.

```js
// helpers/_foo.js

function _foo() {
  // ...
}

export { _foo as _ }
```
```js
// index.js

// We know the func name will be `_foo` based on the filename.
export { _ as _foo } from '_foo.js'
```

Closes #9203

### CODE REVIEW

File: packages/helpers/esm/_array_without_holes.js
Comment (magic-akari) [inline packages/helpers/esm/_array_without_holes.js:1]: All import statements also need to be corrected.

**Review by magic-akari — CHANGES_REQUESTED** [review]

**Review by magic-akari — CHANGES_REQUESTED** [review]
We also need to update code at packages/helpers/scripts/ast_grep.js, ensuring that the generation logic for CJS is correct.

**Comment by devjiwonchoi** [comment]
Oh, yes! Working on it. 😆

**Comment by magic-akari** [comment]
There is some code logic that needs to be referred to at https://ast-grep.github.io. If you need assistance, please let us know.

File: packages/helpers/scripts/ast_grep.js
Comment (devjiwonchoi) [inline packages/helpers/scripts/ast_grep.js:90]: this seemed to need a space, lmk if it is intended - no error thrown yet.

Comment (magic-akari) [reply packages/helpers/scripts/ast_grep.js:90]: This does not matter since ast-grep works on the AST, not based on string matching. 
And inserting a space enhances the visual formatting.

Comment (devjiwonchoi) [reply packages/helpers/scripts/ast_grep.js:90]: Ah, understood.

Comment (magic-akari) [reply packages/helpers/scripts/ast_grep.js:90]: In fact, `export_alias` and the modified `match` have matched the same sentence. 
So I think we can omit one query.

File: packages/helpers/scripts/ast_grep.js
Comment (devjiwonchoi) [inline packages/helpers/scripts/ast_grep.js:66]: @magic-akari Am I good to omit this as well?

```js
task_queue.push(	
  fs.appendFile(tree.filename(), export_shortname, "utf-8"),	
);
```

File: packages/helpers/scripts/ast_grep.js
Comment (devjiwonchoi) [inline packages/helpers/scripts/ast_grep.js:97]: @magic-akari Am I good to omit this as well?

```js
task_queue.push(	
  fs.appendFile(tree.filename(), export_shortname, "utf-8"),	
);
```

Comment (magic-akari) [reply packages/helpers/scripts/ast_grep.js:97]: It's OK.

File: packages/helpers/scripts/ast_grep.js
Comment (magic-akari) [inline packages/helpers/scripts/ast_grep.js:87]:
```suggestion
                    source.prependLeft(range.start.index, `exports._ = `);
```

There is also a similar statement on line 73.

File: packages/helpers/scripts/ast_grep.js
Comment (magic-akari) [inline packages/helpers/scripts/ast_grep.js:69]: You should directly use the end index of the match.

File: packages/helpers/scripts/ast_grep.js
Comment (devjiwonchoi) [reply packages/helpers/scripts/ast_grep.js:87]: I think we can just remove the whole forEach part:

```js
match
  .findAll({
      rule: {
          pattern: func_name,
          kind: "identifier",
          inside: { kind: "assignment_expression", field: "left" },
      },
  })
  .forEach((match) => {
      const range = match.range();
  
      source.prependLeft(range.start.index, `exports._ = exports.${func_name} = `);
  });
```

This is doing the same thing:

```js
source.update(
  export_start,
  export_end,
  `exports._ = exports.${func_name} = ${func_name};`,
);
```

File: packages/helpers/scripts/ast_grep.js
Comment (devjiwonchoi) [reply packages/helpers/scripts/ast_grep.js:69]: This is doing same as:

```js
source.prependLeft(range.start.index, `exports._ = exports.${func_name} = `);
```

If I use `match.range().end.index` it throws as:

```
file:///Users/devjiwonchoi/DevLife/forks/swc/node_modules/magic-string/dist/magic-string.es.mjs:761
                if (end > this.original.length) throw new Error('end is out of bounds');
                                                      ^

Error: end is out of bounds
    at MagicString.update (file:///Users/devjiwonchoi/DevLife/forks/swc/node_modules/magic-string/dist/magic-string.es.mjs:761:41)
    at file:///Users/devjiwonchoi/DevLife/forks/swc/packages/helpers/scripts/ast_grep.js:70:20
```

Comment (devjiwonchoi) [reply packages/helpers/scripts/ast_grep.js:69]: Ah, you're right, sorry.

File: packages/helpers/scripts/ast_grep.js
Comment (devjiwonchoi) [reply packages/helpers/scripts/ast_grep.js]: I think it can also be just:

```js
`exports._ = ${func_name};`
```

Since we are going to access `_` only?

Comment (magic-akari) [reply packages/helpers/scripts/ast_grep.js]: Function declarations are hoisted. But using function expressions directly might cause issues. This is especially true in cases of self-recursion within the function. You need to verify it.

Comment (magic-akari) [reply packages/helpers/scripts/ast_grep.js]: > I think we can just remove the whole forEach part:
> 
> ```js
> match
>   .findAll({
>       rule: {
>           pattern: func_name,
>           kind: "identifier",
>           inside: { kind: "assignment_expression", field: "left" },
>       },
>   })
>   .forEach((match) => {
>       const range = match.range();
>   
>       source.prependLeft(range.start.index, `exports._ = exports.${func_name} = `);
>   });
> ```
> 
> This is doing the same thing:
> 
> ```js
> source.update(
>   export_start,
>   export_end,
>   `exports._ = exports.${func_name} = ${func_name};`,
> );
> ```

This is used to modify the internal references within a function. Consider the following example.

https://github.com/swc-project/swc/blob/1ab406cd7aa19ea333a8462b0cd496ceb3e39ac1/packages/helpers/esm/_construct.js

```JavaScript
import { _is_native_reflect_construct } from "./_is_native_reflect_construct.js";
import { _set_prototype_of } from "./_set_prototype_of.js";
export function _construct(Parent, args, Class) {
    if (_is_native_reflect_construct()) _construct = Reflect.construct;
    else {
        _construct = function construct(Parent, args, Class) {
            var a = [null];
            a.push.apply(a, args);
            var Constructor = Function.bind.apply(Parent, a);
            var instance = new Constructor();

            if (Class) _set_prototype_of(instance, Class.prototype);

            return instance;
        };
    }

    return _construct.apply(null, arguments);
}
export { _construct as _ };
```

Please note that the exported `_construct` will be updated within the function body.

**Review by kdy1 — DISMISSED** [review]
Thank you!

---

swc-bump: 
 - dbg-swc

**Comment by kdy1** [comment]
I'll merge once @magic-akari approves this PR

File: packages/helpers/scripts/ast_grep.js
Comment (devjiwonchoi) [reply packages/helpers/scripts/ast_grep.js]: Cool, had the same fix lol. :)

added comment why we find from `tree.root`!

**Comment by magic-akari** [comment]
There are some helpers from the 'tslib', should they be processed in the same way? @kdy1

**Review by magic-akari — APPROVED** [review]

**Comment by kdy1** [comment]
Those are fine. My point was that extra exports are useless, but creates extra entries for some tree shaking thing in turbopack.

**Comment by devjiwonchoi** [comment]
@magic-akari Thank you for so detailed, kind review!!

**Pull request merged**
